### ISSUE TITLE
Text highlighting: should we preserve density?

### ISSUE THREAD

**Issue by lopuhin** [TeamHG-Memex/eli5#30]
When highlighting a feature, we can highlight it regardless of length (currently in master), or try to preserve density, so coloring longer feature with a less intense color. I tried that second approach in preserve-density branch, here are some screenshots with master behaviour on top (links to notebooks: https://github.com/TeamHG-Memex/eli5/blob/preserve-density/notebooks/explain_text_prediction.ipynb for words and https://github.com/TeamHG-Memex/eli5/blob/preserve-density/notebooks/explain_text_prediction_char.ipynb for chars).

![2016-10-21 11 28 54](https://cloud.githubusercontent.com/assets/424613/19592008/94205cf6-9782-11e6-8ceb-16f1b746fa76.png)

![2016-10-21 11 31 22](https://cloud.githubusercontent.com/assets/424613/19592013/98244ede-9782-11e6-84b5-ce63bd1dfe70.png)

**Comment by kmike** [comment]
I don't see a big difference in these examples. For tokens it made it worse. For chars it probably made it more correct. I think it is a good option to have, +1 to add it and enable for char ngrams.

**Comment by kmike** [comment]
For char_wb and for token ngrams it is not clear what is better.

**Comment by lopuhin** [comment]
I agree that it's worse for tokens, but even for char ngrams, the main difference seems to be that bias (the largest feature) is less intense compared to shorter char n-grams, so they look brighter (about the same brightness as bias).
I'd rather have it off by default in all cases, but I'm not entirely sure.
One more questions: I just realised that currently weights that are displayed on mouse over in title are different depending on the preserve_density option value, that is clearly wrong, we should have original weights in title, right?

**Comment by kmike** [comment]
Let's take 'software' token in the following text as an example:

> <BIAS> hi there, i am here looking for some help. my friend is a interior decor designer. he is from thailand. he is trying to find some graphics software on pc. any suggestion on which software to buy,where to buy and how much it costs ? he likes the most sophisticated software(the more features it has,the better)

In char-based example it is light-green. The total weight it gives to the result is quite large though: about 0.1 (a bit less for some of them) for each of char 3- and 4-grams (' sof', 'sof', 'soft', 'oft', 'oftw', 'ftw', 'ftwa', 'twa', 'twar', 'war', 'ware', 'are'); it sums up to a value larger than BIAS. But bias is much brighter here:

![2016-10-21 16 55 54](https://cloud.githubusercontent.com/assets/107893/19602902/4a4c0954-97af-11e6-9fb4-98350b2527d0.png)

It has L value which is 2x brighter than the brightest part of 'software'. In token-based example 'software' is bright-green because it gets a single large weight.

**Comment by lopuhin** [comment]
Indeed, you are right, I see the point now!

**Comment by lopuhin** [comment]
> One more questions: I just realised that currently weights that are displayed on mouse over in title are different depending on the preserve_density option value, that is clearly wrong, we should have original weights in title, right?
After thinking about it more, I think we should display the same weight in title as we use for highlighting, especially in the light of your last example.

**Comment by lopuhin** [comment]
Fixed in #32
